refactor: Move chain constants to the util library

sedited opened the pull request:
This pull request is part of the `libbitcoinkernel` project https://github.com/bitcoin/bitcoin/issues/24303 https://github.com/bitcoin/bitcoin/projects/18 and more specifically its "Step 2: Decouple most non-consensus code from libbitcoinkernel". It is also a follow up to #26177.

It replaces pull request https://github.com/bitcoin/bitcoin/pull/27294, which just moved the constants to a new file, but did not re-declare them as enums.

The code move of the chain name constants out of the `chainparamsbase` to their own separate header allows the kernel `chainparams` to no longer include `chainparamsbase`. The `chainparamsbase` contain references to the `ArgsManager` and networking related options that should not belong to the kernel library. Besides this move, the constants are re-declared as enums with helper functions facilitating string conversions.

sedited committed:
refactor: Create chaintype files

This is the first of a number of commits with the goal of moving the
chain type definitions out of chainparamsbase to their own file and
implementing them as enums instead of constant strings. The goal is to
allow the kernel chainparams to no longer include chainparamsbase.

The commit is part of an ongoing effort to decouple the libbitcoinkernel
library from the ArgsManager and other functionality that should not be
part of the kernel library.

sedited committed:
refactor: Replace string chain name constants with ChainTypes

This commit effectively moves the definition of these constants
out of the chainparamsbase to their own file.

Using the ChainType enums provides better type safety compared to
passing around strings.

The commit is part of an ongoing effort to decouple the libbitcoinkernel
library from the ArgsManager and other functionality that should not be
part of the kernel library.

sedited committed:
Add missing definitions in prep for scripted diff

The missing include and ArgsManager were found after applying the
scripted diff in the following commit.

sedited committed:
scripted-diff: Remove unused chainparamsbase includes

This is a follow-up to previous commits moving the chain constants out
of chainparamsbase.

The script removes the chainparamsbase header in all files where it is
included, but not used. This is done by filtering against all defined
symbols of the header as well as its respective .cpp file.

The kernel chainparams now no longer relies on chainparamsbase.

-BEGIN VERIFY SCRIPT-
sed -i '/#include <chainparamsbase.h>/d' $( git grep -l 'chainparamsbase.h' | xargs grep -L 'CBaseChainParams\|CreateBaseChainParams\|SetupChainParamsBaseOptions\|BaseParams\|SelectBaseParams\|chainparamsbase.cpp' )
-END VERIFY SCRIPT-

fanquake commented:
```bash
7ea09dd60  scripted-diff: Remove unused chainparamsbase includes
d99bd93e0 Add missing definitions in prep for scripted diff
5589a4b0d refactor: Replace string chain name constants with ChainTypes
bd6493bf3 refactor: Create chaintype files

Error: script block marker but no scripted-diff in title of commit 7ea09dd60cd43e943f9910c950b05dda89ef3725
Failed
```

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/27491#issuecomment-1514492185

> 7ea09dd60  scripted-diff: Remove unused chainparamsbase includes

Seems like the trailing whitespace in the commit mesasge header was tripping it up?

ryanofsky commented on src/chainparams.cpp:
```diff
@@ -96,26 +97,30 @@ const CChainParams &Params() {
 }
 
-std::unique_ptr<const CChainParams> CreateChainParams(const ArgsManager& args, const std::string& chain)
+std::unique_ptr<const CChainParams> CreateChainParams(const ArgsManager& args, const ChainType& chain)
```
In commit "refactor: Replace string chain name constants with ChainTypes" (5589a4b0dd5b25017299f666127b7537f5f7a4c1)

Maybe replace `const ChainType&` with `ChainType` everywhere since it's an enum and probably more efficient to pass by value instead of by reference

ryanofsky reviewed (approved):
Code review ACK bbe05915adad45560907b5a9b58fc55052838c08. Left a few suggestions, but this looks good in its current form

ajtowns reviewed (commented):
Would be better if the second commit (renaming `CBaseChainParam::MAIN` to `ChainType::MAIN` etc) could be a scripted-diff. Perhaps structure it as:

 * introduce ChainType
 * add `CreateBaseChainParams(ChainType)` and replace `CreateBaseChainParams(string)` with `{ auto ct = ChainTypeFromString(chain); if (ct) { return CreateBaseChainParams(ct); } else { throw...; }` (and so on)
 * scripted-diff to switch to `ChainType::MAIN` and `args.GetChainType()`
 * remove the now unused functions that accepted a string

TheCharlatan commented:
Thank you for the reviews!

Updated bbe05915adad45560907b5a9b58fc55052838c08 -> bbe05915adad45560907b5a9b58fc55052838c08 ([kernelChainType_0](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_0) -> [kernelChainType_1](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_0..kernelChainType_1))
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1171175379), removing `ChainType` pass by reference.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1171598387), removing the switch statement default cases.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1171619487), updating the comment as suggested.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1171633479), cleaning up remaining mentions of the now renamed functions and variable names
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1171681447), adding a commit with the suggested refactor for avoiding needless type conversions.
* Addressed @ajtown's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1172076982), improving enum declaration formatting.

TheCharlatan commented on src/util/chaintype.h:
```diff
@@ -0,0 +1,20 @@
+enum class ChainType { MAIN,
+                       TESTNET,
+                       SIGNET,
+                       REGTEST };
```
The style guide is vague on enums (though I guess you could follow the recommendation for classes)
```
  - Braces on new lines for classes, functions, methods.
  - Braces on the same line for everything else.
```
and our code (even the most recent additions) has a mix between putting braces on the same line and a new line for enums.

TheCharlatan commented on src/util/chaintype.cpp:
```diff
@@ -0,0 +1,40 @@
+#include <stdexcept>
+#include <string>
+
+std::string ChainTypeToString(ChainType chain)
```
I'm not sure on this one. Granted, `std::string ChainName(ChainType chain)` would make it clear for the caller that this is most likely a label associated with the `ChainType`. However, for consistency I would also have to rename the helper functions to `GetChainName()`, and that seems less clear, because it does not have a typed argument allowing the caller to infer the relationship with `ChainType`.

TheCharlatan commented:
Rebased 32a70863215ac14b6402fd2a054aedbfe01ff30d -> 7d361083d72a267de4af258e574219abdef0fc82 ([kernelChainType_1](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_1) -> [kernelChainType_2](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_1..kernelChainType_2))

TheCharlatan commented:
Rebased 7d361083d72a267de4af258e574219abdef0fc82 -> dd95e2a3353b5ded87d1d5408a51bf461f1f90b4 ([kernelChainType_2](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_2) -> [kernelChainType_3](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_2..kernelChainType_3))
* Fixed conflict with #27570

TheCharlatan commented:
Rebased dd95e2a3353b5ded87d1d5408a51bf461f1f90b4 -> 95744f9cf1f143b6449a5046a914557b3886e3a2 ([kernelChainType_3](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_3) -> [kernelChainType_4](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_3..kernelChainType_4))
* Fixed conflict with #17860

ryanofsky commented on src/common/args.cpp:
```diff
@@ -717,14 +718,24 @@ fs::path ArgsManager::GetConfigFilePath() const
 }
 
-std::string ArgsManager::GetChainName() const
+ChainType ArgsManager::GetChainType() const
```
In commit "refactor: Replace string chain name constants with ChainTypes" (ca13faf0cc8acbaf425ac69c783415465403b5a3)

Could this new method be added in an earlier commit? This commit is very long and all the other changes in the commit are mechanical changes except this one method, so the new code seems to get lost.

(This would also open the door to automating the other changes in this commit with a scripted-diff, though that would still be awkward not and not necessary IMO)

ryanofsky commented on src/common/args.h:
```diff
@@ -420,6 +421,8 @@ class ArgsManager
      */
     const fs::path& GetDataDir(bool net_specific) const;
 
+    std::variant<ChainType, std::string> GetChainArg() const;
```
In commit "refactor: Avoid needless chain type string conversions" (95744f9cf1f143b6449a5046a914557b3886e3a2)

Would be helpful to have a code comment describing what this does. Like "Return -regtest/-signet/-testnet/-chain= setting as a ChainType enum if a recognized chain name was set, or as a string if an unrecognized chain name was set. Raise an exception if an invalid combination of flags was provided.

ryanofsky commented on src/common/args.cpp:
```diff
@@ -720,15 +722,19 @@ fs::path ArgsManager::GetConfigFilePath() const
+    return std::get<std::string>(arg);
+}
+
+std::variant<ChainType, std::string> ArgsManager::GetChainArg() const
```
In commit "refactor: Avoid needless chain type string conversions" (95744f9cf1f143b6449a5046a914557b3886e3a2)

It seems like this commit could be moved before the "refactor: Replace string chain name constants with ChainTypes" (ca13faf0cc8acbaf425ac69c783415465403b5a3)" commit, which would simplify that commit, and avoid for the `ArgsManager::GetChainType()` method to change after it is added. I think the only change you would need to make to the code in this commit to move it earlier would be to keep the `ArgsManager::GetChainTypeString()` method called `ArgsManager::GetChainName()`

ryanofsky reviewed (approved):
Code review ACK 95744f9cf1f143b6449a5046a914557b3886e3a2.

I think this looks good in it's current form, but the PR would have less churn and be simpler overall if the last commit "refactor: Avoid needless chain type string conversions" (95744f9cf1f143b6449a5046a914557b3886e3a2) was done earlier and became the second commit before commit "refactor: Replace string chain name constants with ChainTypes" (ca13faf0cc8acbaf425ac69c783415465403b5a3)

sedited committed:
refactor: Introduce ChainType getters for ArgsManager

These are introduced for the next commit where the usage of the
ChainType is adopted throughout the code.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>
Co-authored-by: TheCharlatan <seb.kung@gmail.com>

TheCharlatan commented on src/util/chaintype.cpp:
```diff
@@ -0,0 +1,39 @@
+    case ChainType::REGTEST:
+        return "regtest";
+    }
+    throw std::invalid_argument("Invalid ChainType value");
```
Adding `Assert(false)` seems to throw an error:
```
util/chaintype.cpp: In function ‘std::string ChainTypeToString(ChainType)’:
./util/check.h:73:89: error: control reaches end of non-void function [-Werror=return-type]
   73 | #define Assert(val) inline_assertion_check<true>(val, __FILE__, __LINE__, __func__, #val)
      |                                                                                         ^
util/chaintype.cpp:23:5: note: in expansion of macro ‘Assert’
   23 |     Assert(false);
      |     ^~~~~~
```

TheCharlatan replied:
That works :)

TheCharlatan commented:
Thank you for the reviews!

Updated 95744f9cf1f143b6449a5046a914557b3886e3a2 -> 43dec8806f679005d4484b229f433687526c9133 ([kernelChainType_4](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_4) -> [kernelChainType_5](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_4..kernelChainType_5))
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1187440836), adding another switch case
* Addressed @ryanofsky's [comment_1](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1187490948), [comment_2](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1187514090), re-ordering the commits and declaring the new ChainType getters before using them in the following commits
* Addressed @ryanofsky's  [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1187504635), adding the proposed comment to `GetChainArg` in the `common/args.h` header
* Improved `GetChainType` docstring in the `common/args.h`header
* Removed unneeded `tuple` include in `common/args.cpp`

TheCharlatan commented:
Updated 43dec8806f679005d4484b229f433687526c9133 -> 1ae4d9a37b46a5e69ed79066408a6002c2cc0f73 ([kernelChainType_5](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_5) -> [kernelChainType_6](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_5..kernelChainType_6))

* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/27491#discussion_r1188242770), changed throwing an exception to `assert(false)`.

ryanofsky reviewed (approved):
Code review ACK 1ae4d9a37b46a5e69ed79066408a6002c2cc0f73, but I would suggest another change to make second and third commits more reviewable (see below).

re: https://github.com/bitcoin/bitcoin/pull/27491#issuecomment-1539094705

> re-ordering the commits and declaring the new ChainType getters before using them in the following commits

Thanks for the update, but this sort of does the opposite of what I was trying to suggest. I'm trying move substantive code changes *out* of the giant 3rd commit "Replace string chain name constants with ChainTypes", into smaller earlier commits that are more understandable. I think the giant commit should only consist of mechanical replacements and not contain new code.

Here's a branch with what I was suggesting: baa45381e64261c308e90071d8f3d85cdedb4b3a ([branch](https://github.com/ryanofsky/bitcoin/commits/review.27491.4-edit)). The final change is identical to what you have now, I just rearranged the second and third commits so both are smaller and mechanical changes aren't mixed with substantive changes.

TheCharlatan commented:
Thank you for your patience @ryanofsky 

Updated 1ae4d9a37b46a5e69ed79066408a6002c2cc0f73 -> d168458d1ff987e0d741c75ac1d4b63ae0cfb7e7 ([kernelChainType_6](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_6) -> [kernelChainType_7](https://github.com/TheCharlatan/bitcoin/tree/kernelChainType_7), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelChainType_6..kernelChainType_7))

* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27491#pullrequestreview-1418615677), changed commits as suggested. No net changes to the code, but should result in fewer touched lines in the history.

ryanofsky reviewed (approved):
Code review ACK d168458d1ff987e0d741c75ac1d4b63ae0cfb7e7. Just suggested changes since last review. 

If you can, it would also be helpful to mark previous review comments as resolved (github won't let me do this)

> Thank you for your patience

Likewise! Thank you for all the improvements

fanquake commented:
I'm going to merge this shortly. #27125 (which conflicts) will be rebased (some other conflicting changes may also be merged in the interim), along with its current set of review feedback addressed.

fanquake commented on src/chainparams.cpp:
```diff
@@ -97,26 +99,29 @@ const CChainParams &Params() {
     }
-    throw std::runtime_error(strprintf("%s: Unknown chain %s.", __func__, chain));
+    }
+    throw std::invalid_argument(strprintf("%s: Invalid ChainType value", __func__));
```
Followed up in #27611.

fanquake commented on src/chainparamsbase.cpp:
```diff
@@ -40,22 +36,23 @@ const CBaseChainParams& BaseParams()
         return std::make_unique<CBaseChainParams>("regtest", 18443, 18445);
     }
-    throw std::runtime_error(strprintf("%s: Unknown chain %s.", __func__, chain));
+    throw std::invalid_argument(strprintf("%s: Invalid ChainType value", __func__));
```
Followed up in #27611.
